### PR TITLE
fix: use standard DataLoader for validation to release workers and memory

### LINKED ISSUE #25827: Validation retains InfiniteDataLoader workers and prefetched batches between epochs
## GPU comparison — 2026-09-06

[Fresh measurements on PR #26014](https://github.com/ultralytics/ultralytics/pull/26014#issuecomment-5555193776) confirm that the current finite/persistent design drains validation queues. The 1920px small-data check retained about 64 MiB less shared memory with identical predictions. Matched 20-epoch training took 4.8% longer on COCO8 and 11.7% longer on COCO128, with identical per-epoch metrics. Recommend no default loader change from these results alone; the original GB10 memory-pressure workload remains unverified.

## Current status — 2026-09-05

- [x] Lower validation prefetch shipped in #26057 (8.4.139): [current `build_dataloader`](https://github.com/ultralytics/ultralytics/blob/5eda95e839a9601fc479f1b5e042d48d7e77fa13/ultralytics/data/build.py) uses **2 for unshuffled loaders**, retaining 4 for shuffled training. The original report below describes 8.4.120 and prefetch 4; its 80-batch/~27-GiB queue estimate is historical (at depth 2, the same raw-image estimate is 40 batches/~13.2 GiB, before overhead).
- [ ] The reported GB10/1920px/batch16 memory-pressure case is **not yet reproduced on current main**. This session's real-image Linux/macOS worker check confirms workers remain alive after a pass and terminate at `close()`, with prefetch 2. It does not reproduce the original 95–100-GB pressure.
- [ ] #26014 remains open. Its current proposal is finite validation iteration with **persistent workers**, not unconditional worker teardown. Validate its current head against the current prefetch-2 baseline on the reported GB10 workload, measuring peak/settled PSS, epoch wall time, and unchanged metrics together.

Preserve the explicit maintainer decision to reuse validation workers during training. Historical matched measurements of the earlier #26014 teardown version showed unconditional teardown/recreation increased wall time ~25.3% and reduced throughput ~20.2%; **recommend no action on that rejected approach**. Later #26014 historical GPU trials were workload-dependent (COCO8 ~5.7% faster, COCO128 ~3.6% slower), not a universal speedup. Reducing `workers` and/or image/batch size is a current memory-pressure workaround. The remaining issue is bounded resource use under the reported configuration, not simply that a persistent worker is alive.

---

### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train, Val

### General Summary

During detection training, validation uses an `InfiniteDataLoader`. Its worker processes and prefetched batches remain alive after a validation pass completes.

Detection validation also requests twice the configured worker count. With large images, this can retain tens of gigabytes of host memory, saturate the CPU with memory reclaim/compaction, and leave the GPU idle between epochs. There is no exception or traceback. The failure mode is excessive resource retention.

### Actual Behavior

For the following training configuration:

* workers=15
* batch=16
* imgsz=1920
* cache=False
* 20 host CPUs
* NVIDIA GB10 with unified memory
    I observed:
* 15 training pt_data_worker processes
* 20 validation workers, because validation requests workers * 2, capped by the host CPU count
* Approximately 95–100 GB total memory usage
* 100% CPU utilization after validation completed
* kcompactd0 consuming a full CPU core
* GPU mostly idle during the resulting memory pressure
* Validation worker processes still alive after validation exited

Memory Breakdown

The training parent process had approximately:

* 38.6 GiB PSS
* Approximately 36.8 GiB shared-memory PSS
* Approximately 29.8 GiB NVIDIA allocation
    Representative training workers consumed approximately 1.7 GiB PSS each.
    This occurred with cache=False, so the memory usage was not caused by dataset caching.

### Validation Prexetch Memory

At a validation batch size of 32, a raw uint8 image batch at 1920×1920 is approximately 337.5 MiB per batch.
With 20 workers and a prefetch factor of 4, validation may queue as many as 20 workers × 4 batches/worker = 80 batches.

This corresponds to approximately 80 × 337.5 MiB ≈ 27 GiB of raw image data alone, before accounting for multiprocessing, shared-memory, tensor, and Python overhead.
The combination of persistent validation workers and aggressive prefetching therefore creates substantial retained memory pressure even after the validation pass has completed.

### Expected Behavior

* Training may continue using InfiniteDataLoader and persistent workers.
* Validation should use a normal finite `torch.utils.data.DataLoader`.
* Validation workers and prefetch queues should terminate after validation.
* workers=N should not unexpectedly create 2N validation workers.
* Validation should use a conservative prefetch factor to avoid large queued batches when training with high-resolution images.

### Environment

 ```text
  Ultralytics 8.4.120
  Python 3.12.13
  PyTorch 2.13.0+cu130
  Torchvision 0.28.0
  CUDA 13.0

  OS          Linux 6.17.0-1029-nvidia, aarch64
  CPU count   20
  RAM         121.7 GB
  GPU         NVIDIA GB10, 124610 MiB
  GPU count   1

  numpy                 1.26.4
  opencv-python         4.11.0.86
  pillow                12.3.0
  psutil                7.2.2
  polars                1.43.2
  ultralytics-thop      2.1.6
  ultralytics-platform  0.1.5

### Minimal Reproducible Example

  ```python
  from ultralytics.data.build import build_dataloader

  loader = build_dataloader(
      range(64),
      batch=8,
      workers=2,
      shuffle=False,
      device="cpu",
  )

  workers = list(loader.iterator._workers)

  try:
      print("loader type:", type(loader).__name__)
      list(loader)

      alive = [worker.is_alive() for worker in workers]
      print("workers alive after full iteration:", alive)
      print("worker PIDs:", [worker.pid for worker in workers])

      assert not any(alive), "Workers remain alive after the finite pass"
  finally:
      loader.close()
```

### Additional

_No response_

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by Amarnath10i**
## Summary

Fixes #25827

Validation was using `InfiniteDataLoader`, which keeps worker processes and prefetched batches alive between epochs. With large images and high worker counts this retained tens of gigabytes of memory and saturated the CPU after validation completed.

## Changes

- Add `mode` parameter to `build_dataloader`: `'train'` returns `InfiniteDataLoader` (unchanged behavior), `'val'` returns a standard `DataLoader` with `persistent_workers=False` and `prefetch_factor=2`.
- Pass `mode` through from all validation callers (detect, classify, yoloe).
- Add regression test verifying val mode returns plain `DataLoader`.

## Testing

```bash
python -m pytest tests/test_python.py::test_dataloader_val_mode_releases_workers -xvs
python -m pytest tests/test_python.py::test_dataloader_caps_workers_to_batches -xvs
```

Both pass. The new test asserts that:
1. Train mode still returns `InfiniteDataLoader`
2. Val mode returns plain `DataLoader` (not `InfiniteDataLoader`)
3. Val workers have `persistent_workers=False`

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Updated validation dataloaders to use standard PyTorch `DataLoader` instances with non-persistent workers, preventing worker and prefetched-batch retention after validation while preserving `InfiniteDataLoader` for training.

### 📊 Key Changes
- Added a `mode` parameter to `build_dataloader`, supporting training and validation loader behavior.
- Configured validation loaders with `persistent_workers=False` and a prefetch factor of 2.
- Passed validation mode through detection, classification, and YOLOE dataloader callers.
- Added a regression test confirming training uses `InfiniteDataLoader` and validation uses a plain `DataLoader` with non-persistent workers.

### 🎯 Purpose & Impact
Validation workers and prefetched batches can be released after iteration, reducing retained memory and CPU activity between validation passes. Training dataloader behavior remains unchanged.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Amarnath10i, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message confirms your contribution was received, and an Ultralytics engineer will assist with the review. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/data/build.py [hunk 1]
```diff
@@ -369,19 +375,29 @@ def build_dataloader(
...
+        "worker_init_fn": seed_worker,
+        "generator": generator,
+        "drop_last": drop_last,
+    }
+    if pin_memory_device:
+        loader_kwargs["pin_memory_device"] = pin_memory_device
+    if mode == "val":
+        return dataloader.DataLoader(
+            **loader_kwargs,
+            prefetch_factor=2 if nw > 0 else None,
```
Comment (UltralyticsAssistant) [inline ultralytics/data/build.py]: ⚠️ **HIGH**: When validation runs on CPU or MPS, `BaseValidator` forces `workers=0`, so this passes `prefetch_factor=None` to `DataLoader`. PyTorch 1.8 and 1.13 reject an explicit `None` with `num_workers=0`, raising `ValueError` before validation starts; Ultralytics still supports `torch>=1.8.0`. Omit the keyword when `nw == 0` instead of passing `None`. 

**Suggested change:**
```suggestion
            **({"prefetch_factor": 2} if nw > 0 else {}),
```

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The validation loader change is correctly threaded through the affected validation paths, but it introduces a supported-version regression for CPU/MPS validation because of the explicit `prefetch_factor=None`. Fix that compatibility issue; otherwise LGTM.

<details><summary>💬 Posted 1 inline comment</summary>

- ⚠️ **HIGH** `ultralytics/data/build.py:395` When validation runs on CPU or MPS, `BaseValidator` forces `workers=0`, so this passes `prefetch_factor=None` to `DataLoader`. PyTorch 1.8 and 1.13 reject an explicit `None` with `num_workers=0`, raising `ValueError` before validation starts; Ultralytics still supports `torch>=1.8.0`. Omit the keyword when `nw == 0` instead of passing `None`.
</details>

**Comment by Amarnath10i** [comment]
I have read the CLA Document and I sign the CLA

**Comment by Amarnath10i** [comment]
I have applied the suggested change for prefetch_factor and also added a test case to cover the workers=0 (CPU/MPS) validation path to resolve the Codecov coverage gap!

**Comment by pderrenger** [comment]
Thanks for addressing the `prefetch_factor` compatibility issue and adding coverage for the `workers=0` validation path. All current CI checks are passing, so this is ready for continued maintainer review.

**Comment by danielmorozoff** [comment]
Heads up this bug was already brought up and this was identified as intended behavior. #25828 was the proposed patch that showed significant memory allocation improvements but was not accepted and consequently closed.

**Comment by pderrenger** [comment]
Thanks for flagging this. You’re correct that [#25828](https://github.com/ultralytics/ultralytics/pull/25828) was closed because validation-loader reuse during training was considered intentional; this PR makes the same core lifecycle change, so it requires renewed maintainer review against that prior decision before proceeding.

**Comment by glenn-jocher** [comment]
Thanks for the detailed write-up and for signing the CLA, @Amarnath10i.

I'm closing this. It makes the same core lifecycle change as #25828, which was closed on 2026-08-19 with the reasoning that `self.test_loader` is intentionally reused across training epochs, and specifically declining to add a parallel loader-selection mode. This PR adds that mode under the name `mode=` rather than `infinite=`, across the same seven files. Reopening a settled decision needs new evidence against the closure rationale, and I measured in the other direction on both counts.

**The leak does not reproduce.** With `model.val(data="coco128.yaml", workers=4)` on `main`, child processes go 0 → 0 and open FDs 6 → 6 across the call. Two reasons: `BaseValidator` already forces `workers=0` for CPU/MPS standalone val (`ultralytics/engine/validator.py:230-231`), and `Model.val()` retains no validator, so the loader is released on return. Forcing workers through `build_dataloader` directly, `__del__` drops children 9 → 1 (the survivor is the multiprocessing resource-tracker) and FDs 49 → 11. `InfiniteDataLoader.close()` — which `BaseTrainer` already calls on both loaders at `ultralytics/engine/trainer.py:656-658`, added in #25024 — does the same. The cleanup path you're adding already exists at its owner.

**The change costs a lot where it does take effect.** On a spawn platform, with a real coco128 validation dataset at `imgsz=640, batch=16, workers=4`, four repeated validation passes:

|                            | main   | this PR |
| -------------------------- | ------ | ------- |
| steady-state per val pass  | 0.031s | 20.787s |
| construction + 4 passes    | 0.886s | 83.285s |

670x per pass. It reproduces in pure PyTorch with no Ultralytics code (`persistent_workers=False` on spawn, ~5.1s per worker per pass, linear in worker count), and decomposing it shows the time is worker *boot*, not shutdown. Our default is `workers: 8` and detect val doubles the batch, so a default Windows + CUDA run would spend roughly 40 seconds per epoch re-booting validation workers. Setting `persistent_workers=True` collapses it back to 0.001s/pass — but that defeats the PR's stated purpose entirely, which is the tension that makes the current design deliberate.

Green CI is not evidence here: both `BaseTrainer` and `BaseValidator` force `workers=0` on CPU/MPS, so our runners never execute the path this PR changes.

Three smaller notes. The PR also halves validation `prefetch_factor` from 4 to 2, which was already ruled intended behavior on #25827. The new test's `assert val_loader.persistent_workers is False` passes on unmodified `main` — our `InfiniteDataLoader` already reports `persistent_workers=False`, since its persistence comes from holding `self.iterator`, not the torch flag — so it does not actually cover the change. And at +64/−17 the PR is well over our net-added-lines limit and introduces a new default argument to a public API.

What would change the decision: a reproducible defect **outside** the existing ownership paths — workers surviving after `BaseTrainer` teardown, or after a standalone validator is released — with process/FD counts on the affected platform, plus a same-hardware before/after showing no per-epoch validation regression on a spawn platform. On the memory pressure in #25827 specifically, the supported levers today are lowering `workers` or subclassing the trainer; that issue is the right place to keep profiling it.

Genuinely appreciate the effort and the clear write-up — this was a well-presented PR, it just lands on a decision we've already made deliberately.
